### ISSUE TITLE
fail to load after structure member type redefinition

### ISSUE THREAD

**Issue by anj00** [JuliaIO/JLD2.jl#354]
```
julia v1.6.2
JLD2 v0.4.15
Windows 10
```
I have data saved in jld2 files. At some point the module I am using changed the definition of one of a structures. Now JLD2 fails to load the "old" data into the "new" definition. I understand that in general it could be a tough ask. however in this case the puzzling bit is that a simplified (yet having actual modification) version of code works just fine.

This "simple" code actually works
```
using JLD2
struct MyFixedString
    name::String
end
x = MyFixedString("test")

save("struct.jld2", Dict("x" => x))

#restart julia here
using JLD2
using InlineStrings: InlineString15
struct MyFixedString
    name::InlineString15
end

y = load("struct.jld2")["x"]
```

However, once the "modified" structure is used as a member type of a second structure jld2 fails to do the conversion with the following 

```
using JLD2
struct MyFixedString
    name::String
end
struct MyStuct
    myFixedString::MyFixedString
end

x = MyStuct(MyFixedString("test"))
save("struct.jld2", Dict("x" => x))

#restart julia here
using JLD2
using InlineStrings: InlineString15
struct MyFixedString
    name::InlineString15
end
struct MyStuct
    myFixedString::MyFixedString
end
y = load("struct.jld2")["x"]

```
Here is the error
```
Fatal error:
ERROR: LoadError: TypeError: in <:, expected Type, got a value of type JLD2.OnDiskRepresentation{(0,), Tuple{String}, Tuple{JLD2.Vlen{String}}}
```
This is happening with a real module: TimeZones since release 1.6.0, rendering bunch of my historical date not loadable :)  . I just boiled down the problem to mock code size.

**Comment by JonasIsensee** [comment]
Hi @anj00 ,
I stand corrected, this was intended behaviour. Your example should work, using #356

**Comment by anj00** [comment]
Thanks, 
In this particular case, debugging this problem I realized that storing ZonedDateTime is VERY expensive in JLD2 (both HDD size and time it takes to load/save). And that coupled with with risk of ZonedDateTime might change in the future again forced me to rethink the strategy. In my case I realized that It is possible to save DateTime instead of ZonedDateTime (converting to UTC). (Betting on DateTime hopefully less likely to change). So I converted old data (using old version of ZonedDateTime) do DateTime. And changed my algorithm accordingly.  

YMMW, in my case .jld2 file became half the size. and overall processing algo (loading, processing, saving) became 4x faster.  

I will take a look on #242 next time I run into similar problem. 

I am not in the context. You probably discussed it many times in other places. This situation with modules changing definitions of struct and JLD2 not being able to read the old data from this point on is very annoying and costing tremendous amounts of time to fix every time a module does a change. (In my case every time that happens I have to write a chunk of custom conversion code.). So if I have a jld2 written 2-3 years back I am pretty sure I won't be able to read it today (unless I am "lucky" enough to have only very basic types there or non changed structs) 
 
I have close to zero experience here, just first thing which pops up:
Ideally of course, one should have some sort of versioning. so each module has conversion functions from many old versions to the current. JLD2 would then apply that function (giving a warning to a user, so user can potentially fix the file).  Essentially shifting the problem of writing the conversion code from many thousands of users to a module developer. But I bet this is not as simple and an expert would shoot 10 holes in this idea in a blink of an eye :)
